core/governance/market.go

```diff
@@ -832,3 +957,89 @@ func validateUpdateFuture(future *types.UpdateFutureProduct, et *enactmentTime)
 
 	return types.ProposalErrorUnspecified, nil
 }
+
+func validateUpdatePerps(perps *types.UpdatePerpsProduct, et *enactmentTime, currentTime time.Time) (types.ProposalError, error) {
+	settlData := &perps.DataSourceSpecForSettlementData
+	if settlData == nil {
+		return types.ProposalErrorInvalidPerpsProduct, ErrMissingDataSourceSpecForSettlementData
+	}
+
+	if settlData.Content() == nil {
+		return types.ProposalErrorInvalidPerpsProduct, ErrMissingDataSourceSpecForSettlementData
+	}
+
+	ext, err := settlData.IsExternal()
+	if err != nil {
+		return types.ProposalErrorInvalidPerpsProduct, err
+	}
+
+	if !ext {
+		return types.ProposalErrorInvalidPerpsProduct, ErrSettlementWithInternalDataSourceIsNotAllowed
+	}
+
+	settlSchedule := &perps.DataSourceSpecForSettlementSchedule
+	if settlSchedule == nil {
+		return types.ProposalErrorInvalidPerpsProduct, ErrMissingDataSourceSpecForSettlementSchedule
+	}
+
+	if settlSchedule.Content() == nil {
+		return types.ProposalErrorInvalidPerpsProduct, ErrMissingDataSourceSpecForSettlementSchedule
+	}
+
+	if perps.DataSourceSpecBinding == nil {
+		return types.ProposalErrorInvalidPerpsProduct, ErrMissingDataSourceSpecBinding
+	}
+
+	// ensure the oracle spec for settlement data can be constructed
+	ospec, err := spec.New(*datasource.SpecFromDefinition(perps.DataSourceSpecForSettlementData))
+	if err != nil {
+		return types.ProposalErrorInvalidPerpsProduct, err
+	}
+	switch perps.DataSourceSpecBinding.SettlementDataProperty {
+	case datapb.PropertyKey_TYPE_DECIMAL.String():
+		err := ospec.EnsureBoundableProperty(perps.DataSourceSpecBinding.SettlementDataProperty, datapb.PropertyKey_TYPE_DECIMAL)
+		if err != nil {
+			return types.ProposalErrorInvalidPerpsProduct, fmt.Errorf("invalid oracle spec binding for settlement data: %w", err)
+		}
+	default:
+		err := ospec.EnsureBoundableProperty(perps.DataSourceSpecBinding.SettlementDataProperty, datapb.PropertyKey_TYPE_INTEGER)
+		if err != nil {
+			return types.ProposalErrorInvalidPerpsProduct, fmt.Errorf("invalid oracle spec binding for settlement data: %w", err)
+		}
+	}
+
+	// ensure the oracle spec for market termination can be constructed
+	ospec, err = spec.New(*datasource.SpecFromDefinition(perps.DataSourceSpecForSettlementSchedule))
+	if err != nil {
+		return types.ProposalErrorInvalidPerpsProduct, err
+	}
+
+	switch perps.DataSourceSpecBinding.SettlementScheduleProperty {
+	case spec.BuiltinTimeTrigger:
+		tt := perps.DataSourceSpecForSettlementSchedule.GetInternalTimeTriggerSpecConfiguration()
+		if len(tt.Triggers) != 1 {
```

Comment:
But tt.Trigers is an array with length 1, do we need this check really?

```var nums[5] int``` outputs ```[0 0 0 0 0]```

Reply:
You are correct. Just did that as for now this is pretty much meant to be an implementation detail, and the hard requirement is that we don’t want multiple of them whatever the data structure used down the line.